src/dotnet/modules/DiffSharpModule/gmm.fs

```diff
@@ -0,0 +1,94 @@
+﻿module gmm
+
+open DiffSharp.AD.Float64
+open DotnetRunner.Data
+open System.Composition
+open DiffSharp.Util
+
+let unpackQ (logdiag: DV) (lt: DV): DM =
+    let d = logdiag.Length
+    DM.init d d (fun i j ->
+                    if i < j then D 0.
+                    else if i = j then exp logdiag.[i]
+                    else lt.[d * j + i - j - 1 - j * (j + 1) / 2])
+
+let logGammaDistrib a p = 0.25 * float(p) * float(p - 1) * log System.Math.PI + ([1..p] |> List.sumBy (fun j -> MathNet.Numerics.SpecialFunctions.GammaLn (a + 0.5 * float(1 - j))))
+
+let logWishartPrior (qsAndSums: (DM * D) array) wishartGamma wishartM p =
+    let k = qsAndSums.Length
+    let n = p + wishartM + 1
+    let c = float(n * p) * (log wishartGamma - 0.5 * log 2.) - (logGammaDistrib (0.5 * float n) p)
+    let frobenius = qsAndSums |> Array.sumBy (fun qAndSum ->
+        let q = fst qAndSum
+        q.GetRows () |> Seq.sumBy DV.l2normSq)
+
+    let sumQs = qsAndSums |> Array.sumBy snd
+
+    0.5 * wishartGamma * wishartGamma * frobenius - float wishartM * sumQs - float k * c
+
+let gmmObjective (alphas: DV) (means: DM) (icf: DM) (x: DM) (wishartGamma: float) (wishartM: int) =
+    let d = x.Cols
+    let n = x.Rows
+    let constant = - float n * float d * 0.5 * log (2. * System.Math.PI)
+    let alphasAndMeans = Seq.zip (alphas.ToArray ()) (means.GetRows ()) |> Array.ofSeq
```

Comment:
Why do we need ``alphas.ToArray()`` Presumably DV doesn't implement IEnumerable?  
Would it be performant to pass means as a ``DV[]`` and icf as a ``DV[]``.  That's actually closer to the mathematics.

Reply:
Yes, DV is not enumerable.
Passing means and icfs as `DV array` instead of `DM` had zero impact on performance, but made the code a bit prettier.